### ISSUE TITLE
Per-toot language

### ISSUE THREAD

**Issue by Wuzzy2** [mastodon/mastodon#17073]
### Pitch

I propose that the language can be specified for individual toots. By default, the language is whatever you specified in your profile as your "toot language" but you can choose to select a different langauge for individual toots.

### Motivation

For example, when you mostly toot in English, but occasionally toot in German as well. Currently, all your toots are marked as being in the language you set in the profile.

**Comment by Gargron** [comment]
In the API, `language` can be specified for an individual post. This is not exposed in the web UI to avoid cluttering the form.

**Comment by rriemann-eu** [comment]
Can we please make this language GUI setting exposure configurable? This would be relevant for accounts that reach out to an audience with high language diversity.

**Comment by Wuzzy2** [comment]
I agree with rriemann-eu. This is exactly the use-case I was aiming at here. :-)

**Comment by piegamesde** [comment]
Note that the toot language is especially important for screen readers, so that they know how to pronounce things.

Would a simple auto-detection algorithm possible that then prompts the user to change the language? I fear that simply having it as an option won't be enough because users might forget to change the setting.

**Comment by Cassolotl** [comment]
This is a duplicate of #8753, but that was closed over three years ago, so I feel like it would be okay to leave this open. :D

**Comment by Manndung75** [comment]
Hanoi

**Comment by gunchleoc** [comment]
I'd like this feature too. I am trilingual.

**Comment by LoveIsGrief** [comment]
I like @gunchleoc's proposal. Maybe the language selection on the web UI could just be a flag of the language and clicking on it opens a dropdown or a modal dialog.

![afbeelding](https://user-images.githubusercontent.com/2829538/165509763-82f25ba8-c518-4fbd-b186-0d5b9e9b4095.png)

**Comment by gunchleoc** [comment]
I have downvoted the flag option, because it won't work. Many countries are multilingual, e.g. does the Scottish flag stand for Gaelic or Scots? And the Nigerian flag for Yorùbá, Ẹ̀dó or any other of the myriad of languages spoken in that country? Which flag would you use for speaker of languages that are spread over multiple nation states, like Ewe (Ghana, Togo, Benin)?

**Comment by Wuzzy2** [comment]
Most languages, with a very few exceptions (e.g. Esperanto, Lojban, Ido), don’t actually have flags. Using national flags for languages doesn’t actually work, because there is not an 1:1 relationship of nation to language. It’s actually an n:m relationship. Therefore, whatever national flag you use to represent a language, this will be completely arbitrary and should therefore not be done.

Just use the ISO language code instead (2 or 3 letters).

**Comment by Oblomov** [comment]
See also #9550. It's honestly a bit impractical that one needs to use external tools to be able to set the toot language when posting. If cluttering the UI is that big of an issue, making the selection box opt-in from the preferences (Preferences > Other > Posting default > add a checkbox 'show language selection in composer' or something) should meet the needs of both those that want a simpler interface and/or post always in the same language, and the others.

**Comment by mro** [comment]
the API requires an ISO language code. Why not just expose it. Verbatim. We are used to codes with international phone numbers (no flags here, right?).

It's brief, concise, human readable, requires no mapping and hides nothing.

**Comment by mro** [comment]
@Gargron 
> avoid cluttering the form.

If the standard is cluttered the UI can't help it if implementing the standard.

**Comment by dahlia** [comment]
> If cluttering the UI is that big of an issue, making the selection box opt-in from the preferences (Preferences > Other > Posting default > add a checkbox 'show language selection in composer' or something) should meet the needs of both those that want a simpler interface and/or post always in the same language, and the others.

Second this.  Few more thoughts and suggestions on UI things:

- Exposing ISO language codes wouldn't be the best experience for average users.  Quite many codes are even unpredictable.  Instead, we could use native language names compiled in Unicode CLDR.

- In addition to @Oblomov's idea, we can further clutter the UI by showing only user's *filter languages* (which can be configured from *Preferences* → *Other* → *Filter languages*).

**Comment by mro** [comment]
> Quite many codes are even unpredictable

same holds for telephone prefixes as I said. Or would you guess 0043 for switzerland? But nevertheless it works perfect and is not an obstacle for international calls.

You won't guess them, but you won't forget your preferred ones either. Users are not morons.

Don't speculate but test it and see people remember the languages they care about.

> only user's filter languages

please don't speculate on limiting decisions. Sort the used ones first, then filter then rest, allow text entry.

And as Stephen Krug says: Don't argue but test it!

**Comment by piegamesde** [comment]
> same holds for telephone prefixes as I said. Or would you guess 0043 for switzerland? But nevertheless it works perfect and is not an obstacle for international calls.

This is a bad argument IMO. Just because it works doesn't mean it's good. The proposal of using the language name instead of its code is a strict improvement usability wise IMO.

**Comment by LoveIsGrief** [comment]
Good arguments were made against flags and I do agree (despite having suggested them at first). The language code might be pretty easy. The web UI could just have the two letter language code underneath the smiley (I'm obviously not a designer):

![afbeelding](https://user-images.githubusercontent.com/2829538/165836051-1ebb1fdd-d169-4b01-aa21-f8b782ea08e5.png)

A click on it could bring up a modal similar to this

![Modal drawio](https://user-images.githubusercontent.com/2829538/165840258-7763bf0d-294f-4412-b51a-4fe4014b462c.png)

**Comment by mro** [comment]
@LoveIsGrief please stick with ISO-639 and make it lowercase. (Upercase is countries, ISO-3166)

**Comment by virgilinojuca** [comment]
I think it should show on the button the ISO code for the selected language, but the full name on the selection window. Not everyone knows which language corresponds to each code.

**Comment by rwmpelstilzchen** [comment]
Is it possible to adapt the [language selection interface used on Wikipedia](https://wikimediafoundation.org/news/2018/03/08/compact-language-links-launch/)? It’s quite neat, powerful and intuitive.

@amire80, any thoughts?

**Comment by eishiya** [comment]
It seems like a lot of users neglect to set their posting language in the first place, or leave it set to English because they post in English sometimes, resulting in the language filters being pretty useless - a lot of (most?) non-English posts are marked as "English", which means if I filter by English, I get a lot of non-English posts, and if I filter by other languages and exclude English, I won't see many relevant posts.

Having a per-toot language option would help draw attention to the fact that the posting language selection even exists, which would make filtering more useful. I'd personally like to see it implemented somewhat like the emoji palette, where it shows all the options, but shows your frequently-used ones at the top, so you can easily select languages you've previously posted in. Users should also have the option to opt out of the selector in their settings since monoglots aren't likely to need it, but doing so should ask them to verify their default posting language. (Incidentally, a similar opt-out for the emote palette might be nice, for those of us who only use a few different emotes and type the codes anyway.)

**Comment by mro** [comment]
thank you for adding this, just saw it at https://digitalcourage.social!

Sadly you chose against languages ISO-639 (lowercase) but for countries ISO-3166 (uppercase). Maybe you find compliant lowercase too scary to bother with.

Nevertheless, having the option to choose is great. Thanks you!

**Comment by Wuzzy2** [comment]
Using the country code instead of language code was definitely a mistake. Languages are not countries, the relationship between country and language is very messy. I hope it will be fixed in a later PR.

**Comment by Gargron** [comment]
Nobody is using country codes.

**Comment by mro** [comment]
@Gargron you may not be aware of it. The according ISO standards are case-sensitive. I see uppercase letters which is ISO-3166 which is countries. Languages(ISO-637) are lowercase.

**Comment by mro** [comment]
I just learned in the linked issue ISO itself is undecided and doesn't make lowercase mandatory. So it be good then.
